### ISSUE TITLE
PodTopologySpread DoNotSchedule-to-ScheduleAnyway fallback mode

### ISSUE THREAD

**Issue by ahg-g** [kubernetes/kubernetes#105977]
### What would you like to be added?

A mode of operation for PodTopologySpread that allows users to express the flexibility to fallback from DoNotSchedule to ScheduleAnyway after a specific condition is met (e.g., after a defined period of time).

### Why is this needed?

This is needed to force the cluster autoscaler to bring up new nodes in under scaled zones, but fallback to schedule anyway if CA can't because of zonal failure or more commonly stockouts.

/sig scheduling

**Comment by alculquicondor** [comment]
What about having a new PodConditionType that the autoscaler can populate when it tried to scale up and failed?

Then the scheduler waits for this condition to be present in the Pod to use the fallback mode.

@liggitt would this be acceptable from an API perspective?

**Comment by liggitt** [comment]
does cluster autoscaler currently write to pod status? do we want it fanning out writes that way (both from the perspective of performing lots of writes and for permissions purposes)?

does scheduler have any awareness of cluster autoscaler today?

**Comment by alculquicondor** [comment]
Cluster Autoscaler currently only writes events. So if CA only updates the pod condition when the condition changes, there is not much risk of too many writes. From a permissions perspective, it would just need write access to the pod status (the same permission that scheduler has). 

Going a bit into technical details, I think this merge strategy allows Pod status updates not to bump into each other if we use server-side-apply both in scheduler and cluster autoscaler https://github.com/kubernetes/api/blob/25b7aa9e86de7bba38c35cbe56701d2c1ff207e9/core/v1/types.go#L3790-L380

But otherwise we would have to add `+listType=map`. Is this backwards compatible?

> does scheduler have any awareness of cluster autoscaler today?

Not at all. But that indeed causes some limitations that this new PodConditionType could solve.

**Comment by sanposhiho** [comment]
Providing this feature with a new `whenUnsatisfiable ` like `DoNotSchedule_BestEffort` (This is a tentative name. i couldn't think of a good name.) looks good to me. This can be implemented simply.
If `whenUnsatisfiable` is `DoNotSchedule_BestEffort`, Pod Topology Spread behaves as `DoNotSchedule` basically, but if Pod cannot be scheduled in a defined period of time, then it changes the behavior to `ScheduleAnyway`. In other words, scheduler waits that time for CA to create Node and if CA doesn't create Node in that time, then schedule Pod with `ScheduleAnyway`. 
And it would be nice if users could define the period of time by looking at how long it usually takes the CA to prepare new Nodes.

**Comment by sanposhiho** [comment]
Looks https://github.com/kubernetes/kubernetes/issues/115320 is the actual use case of this feature.

- Two zones are usually available.
- `zone` topology spread with `minDomains:2` and `DoNotSchedule`

Under this situation, when there is a zone failure, no Pods can be scheduled.
The ideal scenario is that CA realizes unschedulable Pods and creates another Node on a different zone, but it depends on how CA (or node pools etc) is configured. If CA cannot create Pods in a different zone, all Pods with this topology spread will get stuck in unschedulable state. 

If CA cannot create new Node for unschedulable Pods, probably we should allow those Pods to be scheduled within one zone.

**Comment by ssanders1449** [comment]
> Providing this feature with a new `whenUnsatisfiable` like `DoNotSchedule_BestEffort` sounds good and simple to me. (`DoNotSchedule_BestEffort` is a tentative name. i couldn't think of a good name.) If `whenUnsatisfiable` is `DoNotSchedule_BestEffort`, Pod Topology Spread behaves as `DoNotSchedule` basically, but if Pod cannot be scheduled in a defined period of time, then it changes the behavior to `ScheduleAnyway`. It means scheduler waits certain time for CA to create Node and it would be nice if users could define the period of time in pluginConfig(scheduler configuration) by looking at how long it usually takes the CA to prepare new Nodes.

This proposal (DoNotSchedule_BestEffort) would work for https://github.com/kubernetes/kubernetes/issues/117634 as long as the amount of time to wait could be set as part of the constraint. In this use case, we only want to wait the few seconds that it would take to preempt a low priority 'headroom' pod and not the amount of time it usually takes for CA to prepare new nodes.

**Comment by sanposhiho** [comment]
I think we've gotten pretty enough usecases to pursue this idea.
- https://github.com/kubernetes/kubernetes/issues/115320
- https://github.com/kubernetes/kubernetes/issues/116978
- https://github.com/kubernetes/kubernetes/issues/117634

From the past discussion, our ideas so far can be roughly divided into two types:
- decide to fallback to ScheduleAnyway based on how long the Pod has been in unschedulable state ([comment](https://github.com/kubernetes/kubernetes/issues/105977#issuecomment-1082992497))
- decide to fallback to ScheduleAnyway based on new `PodConditionType` to realize a response from CA. ([comment](https://github.com/kubernetes/kubernetes/issues/105977#issuecomment-985644457))

I +1 for @alculquicondor's idea now; new `PodConditionType` to receive a response from CA. 
Because the first idea (from the past me) requires the effort to configure the appropriate threshold time based on their cluster's situation; the time to run up new Nodes probably depends on what the ClusterAutoscaler in the cluster is and what situation unprepared behind machines are in. To prepare the way to receive the response from CA works 100% in any cases and looks easier for people since no configuration is required for them.

Or, instead of new `PodConditionType`, can we use the `NotTriggerScaleUp` event that the current CA emits? ([ref](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-events-are-emitted-by-ca))
@alculquicondor cannot we just utilize `NotTriggerScaleUp` event to notice the need to fallback?

**Comment by ssanders1449** [comment]
> From the past discussion, our ideas so far can be roughly divided into two types:
> 
> * decide to fallback to ScheduleAnyway based on how long the Pod has been in unschedulable state ([comment](https://github.com/kubernetes/kubernetes/issues/105977#issuecomment-1082992497))
> * decide to fallback to ScheduleAnyway based on new `PodConditionType` to realize a response from CA. ([comment](https://github.com/kubernetes/kubernetes/issues/105977#issuecomment-985644457))

A third possibility would be to fall back to ScheduleAnyway immediately if there are no lower priority pods that can be preempted. This also would not require any special configuration, but would require a new 'whenUnsatisfiable' value for this specific use case (PreemptOrSchedule)

**Comment by kerthcet** [comment]
One concern is will `PreemptOrSchedule` work peacefully with default preemption plugin.

**Comment by alculquicondor** [comment]
If it doesn't, we fix it :) A feature can affect more than one plugin.

**Comment by sanposhiho** [comment]
/assign

I'd like to work on it for hopefully v1.28, but probably gonna be v1.29.

**Comment by knelasevero** [comment]
This feature is actually interesting to me, as well as some of the use cases listed in this thread. 

@sanposhiho are you going to be the one writing the KEP?

**Comment by sanposhiho** [comment]
I'm gonna write KEP. You can help reviews to refine our idea.

**Comment by grosser** [comment]
FYI internal tool that we built to work around this issue, a bit clunky but maybe someone finds this useful or an inspiration:
- find pending pods that violate their PDB
- mark pod-template-hash in admission hook for ScheduleAnyway conversion + eviction by descheduler
- delete pods
-> PDB no longer violated

**Comment by ShadowJonathan** [comment]
I found `PreemptOrSchedule` from another thread, and it actually fits my use-case:

We have a mixed web-server queue-worker cluster, and we'd like the web-servers to be spread evenly across the nodes at any one time, to make sure that any load that they will absorb, will be evenly distributed. We have quite bursty thundering-herd traffic, where at one time it may have 20r/s, while at another time sustain to 100-200r/s for a few seconds, so spreading the load evenly is important. (We have another component making sure that the per-pod rate of requests doesn't exceed its capacity, so excess requests get queued up elsewhere)

At the same time, there'll be queue worker pods with lower priorities that will be spread similarly across the nodes.

Currently we have a situation where web-servers are unevenly spread, and we're fine with them pre-empting queue worker pods to be rescheduled on another node, if it means the web-server will evenly spread to that node.

Is this currently possible? Would `PreemptOrSchedule` help with this?

**Comment by sanposhiho** [comment]
@ShadowJonathan 
Looks like your usecase is valid for the feature.
Unfortunately, it has not yet been implemented in the first place, but is in the designing phase. 
You can follow the progress in https://github.com/kubernetes/enhancements/pull/4150.

**Comment by alculquicondor** [comment]
cc @dom4ha

**Comment by grapemix** [comment]
In homelab usecase, some special type of instances (like GPU nodes) are not always turn on because of well predicted low usage time period and the cost of electricity. It make sense to fall back to non GPU nodes with CPU decoding when usage is low and/or GPU nodes are not available. So it is good to have a ScheduleAnyway option.

**Comment by sanposhiho** [comment]
I'd like to pursue this one towards the next release cycle.

@gjtempleton @MaciekPytel can you (or assign someone to) review the opening KEP PR from the sig-autoscaling PoV?

**Comment by sanposhiho** [comment]
/remove-lifecycle stale

Still being blocked on the sig-autoscaling side reviewing.

**Comment by sanposhiho** [comment]
/remove-lifecycle stale

Still waiting for the sig-autoscaling side review.

**Comment by sanposhiho** [comment]
We cannot do that because if pods have `ScheduleAnyway`, pods won't be unschedulable in the first place. i.e., even if users want a fallback behavior and put `ScheduleAnyway`, the scheduler could schedule those pods to some nodes just by (mostly) ignoring topology-aware spreading. 

That's the reason we need the scheduler to distinguish this fallback mode from `ScheduleAnyway` and `DoNotSchedule`. 
It has to handle pods with `DoNotSchedule` first, then CA checks if it can create a new node or not, and if not, _then_ the scheduler starts to handle pods as `ScheduleAnyway` to give up a strict spreading, fallback to a best-effort spreading.

**Comment by x13n** [comment]
Right, that makes sense. I think there's a few design options here.
- Using SchedulingGates to prepare the cluster for pods with `ScheduleAnyway`. Such pods could trigger autoscaling and be passed to scheduler only after the nodes come up. Processing would happen in CA->scheduler.
- Introducing this fallback mode of operation to the existing API to explicitly ask Autoscaler to create new nodes. Processing would happen in scheduler, then CA, then again in scheduler.
- Tighter integration between Cluster Autoscaler and kube-scheduler. For instance, if kube-scheduler was aware of potentially available nodes, it could treat them similarly to actual nodes, potentially spreading pods across them. There's a number of sub-options here and it is definitely a larger change, but it would help with other issues when trying to balance autoscaling & scheduling preferences.

**Comment by sanposhiho** [comment]
First of all, in any of your options, we _must_ create a new API because there's users wanting this fallback feature, while, at the same time, there's other users wanting the existing **best-effort** `ScheduleAnyway` behavior, which doesn't trigger CA.

> Using SchedulingGates to prepare the cluster for pods with ScheduleAnyway. Such pods could trigger autoscaling and be passed to scheduler only after the nodes come up. Processing would happen in CA->scheduler.

I don't like this because it means CA has to get an implementation for gating pods with topology spread. What if we want to implement the same fallback mechanism in other scheduling constraints? Or maybe users want to do that with their custom plugins. When doing that, we/they have to change CA as well so that CA can put the scheduling gate onto the appropriate pods.
On the other hand, `TriggeredScaleUp` (proposed at https://github.com/kubernetes/enhancements/pull/4150) allows us to implement it only on the scheduler plugin changes.
At the filter extension point, if plugins find `TriggeredScaleUp`, it can skip handling this pod and handle this pod at the scoring phase instead (= fallback). Any plugins will be able to implement this fallback mechanism after a proposed KEP.

> explicitly ask Autoscaler to create new nodes. Processing would happen in scheduler, then CA, then again in scheduler.

This is just what we have now, right? We have `PodScheduled: false` condition to ask CA to create a new node for the pod, CA creates a new node (or cannot create a new node), and the scheduler retries scheduling the pod after the node creation.
And, the core point here is what if CA cannot create a new node: we have an API to ask CA to create new nodes (`PodScheduled: false`), but we don't have an API to receive the response from CA when the autoscaling fails. And, https://github.com/kubernetes/enhancements/pull/4150 tries to create a new API, `TriggeredScaleUp`.

> Tighter integration between Cluster Autoscaler and kube-scheduler. For instance, if kube-scheduler was aware of potentially available nodes, it could treat them similarly to actual nodes, potentially spreading pods across them. There's a number of sub-options here and it is definitely a larger change, but it would help with other issues when trying to balance autoscaling & scheduling preferences.

You mean something like https://github.com/kubernetes/enhancements/pull/5287#discussion_r2093115985, right? Yes, I love this. But, I don't think it's related to this feature here, essentially. 
We, either way, need a new condition like `TriggeredScaleUp` so that the scheduler can notice CA cannot create a new node for those pending pods.

**Comment by x13n** [comment]
Thanks, I haven't seen the KEP before (or I forgot I did). I like the idea of using pod conditions for this.

> First of all, in any of your options, we _must_ create a new API because there's users wanting this fallback feature, while, at the same time, there's other users wanting the existing **best-effort** `ScheduleAnyway` behavior, which doesn't trigger CA.

If there was an option to have autoscaling-assisted topology spreading, would anyone still want best effort `ScheduleAnyway`? In static (not autoscaled) clusters it should behave the same in both modes, so the only distinction is when autoscaling is already in the picture.

> You mean something like [kubernetes/enhancements#5287 (comment)](https://github.com/kubernetes/enhancements/pull/5287#discussion_r2093115985), right? Yes, I love this. But, I don't think it's related to this feature here, essentially. We, either way, need a new condition like `TriggeredScaleUp` so that the scheduler can notice CA cannot create a new node for those pending pods.
> 
> So, I still much prefer the implementation change described at [kubernetes/enhancements#5287](https://github.com/kubernetes/enhancements/pull/5287).

Agreed kubernetes/enhancements#5287 is a good improvement for now. I think it is compatible with giving scheduler more information about upcoming nodes, so we should do NNN already and then think on how to improve further.

**Comment by ShadowJonathan** [comment]
still relevant

**Comment by sanposhiho** [comment]
Agree it's still relevant, but not sure if we (sig-scheduling) have a bandwidth working on this. We're prioritizing WAS work mainly, and the remaining time is eaten by DRA generally.
